### PR TITLE
[SPARK-31238][SQL] Rebase dates to/from Julian calendar in write/read for ORC datasource

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?

This PR (SPARK-31238) aims the followings.
1. Modified ORC Vectorized Reader, in particular, OrcColumnVector v1.2 and v2.3. After the changes, it uses `DateTimeUtils. rebaseJulianToGregorianDays()` added by https://github.com/apache/spark/pull/27915 . The method performs rebasing days from the hybrid calendar (Julian + Gregorian) to Proleptic Gregorian calendar. It builds a local date in the original calendar, extracts date fields `year`, `month` and `day` from the local date, and builds another local date in the target calendar. After that, it calculates days from the epoch `1970-01-01` for the resulted local date.
2. Introduced rebasing dates while saving ORC files, in particular, I modified `OrcShimUtils. getDateWritable` v1.2 and v2.3, and returned `DaysWritable` instead of Hive's `DateWritable`. The `DaysWritable` class was added by the PR https://github.com/apache/spark/pull/27890 (and fixed by https://github.com/apache/spark/pull/27962). I moved `DaysWritable` from `sql/hive` to `sql/core` to re-use it in ORC datasource.

### Why are the changes needed?
For the backward compatibility with Spark 2.4 and earlier versions. The changes allow users to read dates/timestamps saved by previous version, and get the same result.

### Does this PR introduce any user-facing change?
Yes. Before the changes, loading the date `1200-01-01` saved by Spark 2.4.5 returns the following:
```scala
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-08|
+----------+
```
After the changes
```scala
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-01|
+----------+
```

### How was this patch tested?
- By running `OrcSourceSuite` and `HiveOrcSourceSuite`.
- Add new test `SPARK-31238: compatibility with Spark 2.4 in reading dates` to `OrcSuite` which reads an ORC file saved by Spark 2.4.5 via the commands:
```shell
$ export TZ="America/Los_Angeles"
```
```scala
scala> sql("select cast('1200-01-01' as date) dt").write.mode("overwrite").orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc")
scala> spark.read.orc("/Users/maxim/tmp/before_1582/2_4_5_date_orc").show(false)
+----------+
|dt        |
+----------+
|1200-01-01|
+----------+
```
- Add round trip test `SPARK-31238: rebasing dates in write`. The test `SPARK-31238: compatibility with Spark 2.4 in reading dates` confirms rebasing in read. So, we can check rebasing in write.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
So, does this happen only in vectorized reader, @MaxGekk ?

**Comment by MaxGekk** [comment]
> So, does this happen only in vectorized reader

@dongjoon-hyun Correct, the regular reader uses `DateTimeUtils.fromJavaDate`:
https://github.com/apache/spark/blob/300ec1a74cb14867c22e616e657566d510426331/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala#L111

And the PRs #27807 and #27980 introduced date rebasing in `fromJavaDate()`.

Also, I fixed ORC writer, and added a round trip test for both vectorized and non-vectorized readers.

**Comment by MaxGekk** [comment]
I think the failure doesn't related to my changes: https://github.com/apache/spark/pull/28016#issuecomment-603958162 . I will update the PR desciption shortly, so the PR will be ready for review. /cc @cloud-fan

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@cloud-fan @dongjoon-hyun @bersprockets Please, review this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala [hunk 1]
```diff
@@ -35,11 +35,12 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils.{rebaseGregorianToJulian
  * @param julianDays The number of days since the epoch 1970-01-01 in
  *                   Julian calendar.
  */
-private[hive] class DaysWritable(
+class DaysWritable(
     var gregorianDays: Int,
     var julianDays: Int)
   extends DateWritable {
 
+  def this() = this(0, 0)
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala]: I assume that `gregorianDays` and `julianDays` will be set later via the `set` method.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM (Pending Jenkins with `hive1.2` to make it sure.)
Thank you for swift fix, @MaxGekk !

**Comment by dongjoon-hyun** [comment]
Hi, @MaxGekk . 
Could you fix the compilation failures in `hive-1.2` profile?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala [hunk 2]
```diff
@@ -35,11 +35,12 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils.{rebaseGregorianToJulian
  * @param julianDays The number of days since the epoch 1970-01-01 in
  *                   Julian calendar.
  */
-private[hive] class DaysWritable(
+class DaysWritable(
     var gregorianDays: Int,
     var julianDays: Int)
   extends DateWritable {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala]: This causes a failure in `hive-1.2` profile due to the following.
- `org.apache.hadoop.hive.serde2.io.DateWritable`
- `org.apache.orc.storage.serde2.io.DateWritable`
```
/home/jenkins/workspace/SparkPullRequestBuilder@3/sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala:54: type mismatch;
[error]  found   : org.apache.spark.sql.execution.datasources.DaysWritable
[error]  required: org.apache.orc.storage.serde2.io.DateWritable
[error]         result
[error]         ^
[info] org.apache.spark.sql.execution.datasources.DaysWritable <: org.apache.orc.storage.serde2.io.DateWritable?
[info] false
[error] /home/jenkins/workspace/SparkPullRequestBuilder@3/sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala:57: type mismatch;
[error]  found   : org.apache.spark.sql.execution.datasources.DaysWritable
[error]  required: org.apache.orc.storage.serde2.io.DateWritable
[error]         new DaysWritable(getter.getInt(ordinal))
```

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala]: We probably need to duplicate this class in v1.2 and v2.3 source code.

File: sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java [hunk 3]
```diff
@@ -130,7 +138,13 @@ public short getShort(int rowId) {
 
   @Override
   public int getInt(int rowId) {
-    return (int) longData.vector[getRowIndex(rowId)];
+    int index = getRowIndex(rowId);
+    int value = (int) longData.vector[index];
```
Comment (cloud-fan) [inline sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java]: nit: `int value = (int) longData.vector[getRowIndex(rowId)];`

File: sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java [hunk 4]
```diff
@@ -42,6 +43,7 @@
   private DecimalColumnVector decimalData;
   private TimestampColumnVector timestampData;
```
Comment (cloud-fan) [inline sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java]: does ORC have `DateColumnVector`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 5]
```diff
@@ -32,6 +32,7 @@ import org.apache.spark.sql.AnalysisException
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.util._
+import org.apache.spark.sql.execution.datasources.DaysWritable
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: unnecessary change

File: sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java (hunk 4 above)
Comment (MaxGekk) [reply sql/core/v1.2/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java]: Yes, it does, but:
1. `DateColumnVector` doesn't have a method similar to `asScratchTimestamp` in `TimestampColumnVector`
2. We don't need to build `java.sql.Date` from serialized days to perform rebasing. It is unnecessary overhead.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala (hunk 5 above)
Comment (MaxGekk) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: If I remove it, the build fails with:
```
Error:(1012, 44) not found: type DaysWritable
  private def getDateWritable(value: Any): DaysWritable =
```
I moved `DaysWritable` from `sql/hive` to `sql/core` to reuse it in ORC.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala]: I cloned `DaysWritable` for Hive ORC v1.2, and added a comment to the cloned class.

**Review by dongjoon-hyun — COMMENTED** [review]
Please move `sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DaysWritable.scala` to `sql/core/v2.3/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala`.

File: sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * The class accepts/returns days in Gregorian calendar and rebase them
+ * via conversion to local date in Julian calendar for dates before 1582-10-15
+ * in read/write for backward compatibility with Spark 2.4 and earlier versions.
+ *
+ * This is a clone of `org.apache.spark.sql.execution.datasources.DaysWritable`.
+ * The class is cloned because Hive ORC v1.2 uses different `DateWritable`:
+ *   - v1.2: `org.apache.orc.storage.serde2.io.DateWritable`
+ *   - v2.3 and `HiveInspectors`: `org.apache.hadoop.hive.serde2.io.DateWritable`
```
Comment (dongjoon-hyun) [inline sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: We don't need the above 4 line comments because `v1.2` and `v2.3` folder structure already is designed for that.

File: sql/core/v2.3/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala [hunk 7]
```diff
@@ -26,6 +26,7 @@ import org.apache.hadoop.hive.ql.io.sarg.PredicateLeaf.{Operator => OrcOperator}
 import org.apache.hadoop.hive.serde2.io.{DateWritable, HiveDecimalWritable}
 
 import org.apache.spark.sql.catalyst.expressions.SpecializedGetters
+import org.apache.spark.sql.execution.datasources.DaysWritable
```
Comment (dongjoon-hyun) [inline sql/core/v2.3/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcShimUtils.scala]: After moving `DaysWritable.scala` to here. I guess we don't need this line. Please try to remove.

**Comment by MaxGekk** [comment]
> Please move ... to v2.3

Why? `DaysWritable` is used in sql/hive, in `HiveInspectors`, just in case.

File: sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: If I move it to 2.3, and enable hive-1.2, `HiveInspectors` will use `DaysWritable` from v1.2 which is wrong or I miss something in your proposal?

**Comment by dongjoon-hyun** [comment]
Ur, are you asking why we keep `v1.2` and `v2.3` structure?

File: sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: Got it. Let me check again.

Comment (dongjoon-hyun) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: ? Sorry, why do you move this to 2.3?

Comment (dongjoon-hyun) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: What I ask is keep this AS-IS, and move the other `DaysWritable`.

Comment (dongjoon-hyun) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: I'm checking again from my side.

Comment (MaxGekk) [reply sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/DaysWritable.scala]: This is what I tried https://github.com/MaxGekk/spark/pull/26, and it fails when I enable hive-1.2:
```
build/sbt -Phive-1.2 clean package
```
Feel free to open a PR for this PR if you mean something different.

**Comment by dongjoon-hyun** [comment]
My bad. I overlooked the difference in `HiveInspectors` in `hive-1.2` profile.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @MaxGekk and @cloud-fan .
Merged to master/3.0.

**Comment by dongjoon-hyun** [comment]
Hi, @MaxGekk . 
Unfortunately, this seems to break all Maven build in master/branch-3.0.
Could you take a look at this?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcTest.scala [hunk 8]
```diff
@@ -133,4 +133,9 @@ abstract class OrcTest extends QueryTest with FileBasedDataSourceTest with Befor
         throw new AnalysisException("Can not match OrcTable in the query.")
     }
   }
+
+  protected def readResourceOrcFile(name: String): DataFrame = {
+    val url = Thread.currentThread().getContextClassLoader.getResource(name)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcTest.scala]: This might be the root cause.

**Comment by dongjoon-hyun** [comment]
I made a follow-up.
- https://github.com/apache/spark/pull/28059
